test: [refactor] Use m_rng directly

maflcko opened the pull request:
This is mostly a style-cleanup for the tests' random generation:


1) `g_insecure_rand_ctx` in the tests is problematic, because the name is a leftover when the generator was indeed insecure. However, now the generator is *deterministic*, because the seed is either passed in or printed (c.f. RANDOM_CTX_SEED). Stating that deterministic randomness is insecure in the tests seems redundant at best. Fix it by just using `m_rng` for the name.


2) The global random context has many one-line aliases, such as `InsecureRand32`. This is problematic, because the same line of code may use the context directly and through a wrapper at the same time. For example in net_tests (see below). This inconsistency is harmless, but confusing. Fix it by just removing the one-line aliases.

```
src/test/net_tests.cpp:        auto msg_data_1 = g_insecure_rand_ctx.randbytes<uint8_t>(InsecureRandRange(100000));
````


3) The wrapper for randmoney has the same problem that the same unit test uses the context directly and through a wrapper at the same time. Also, it has a single type of Rng hardcoded. Fix it by accepting any type.

Commit:
scripted-diff: [test] Use g_rng/m_rng directly

-BEGIN VERIFY SCRIPT-

 # Use m_rng in unit test files
 ren() { sed -i "s:\<$1\>:$2:g" $( git grep -l "$1" src/test/*.cpp src/wallet/test/*.cpp src/test/util/setup_common.cpp ) ; }
 ren InsecureRand32                m_rng.rand32
 ren InsecureRand256               m_rng.rand256
 ren InsecureRandBits              m_rng.randbits
 ren InsecureRandRange             m_rng.randrange
 ren InsecureRandBool              m_rng.randbool
 ren g_insecure_rand_ctx           m_rng
 ren g_insecure_rand_ctx_temp_path g_rng_temp_path

-END VERIFY SCRIPT-

Commit:
test: refactor: Make unsigned promotion explicit

Integer promotion will already turn the `signed` into `unsigned` in
those lines. However, make the `unsigned` explicit so that the code is
clearer and a compiler warning is avoided when switching to m_rng:

| test/validation_block_tests.cpp: warning: comparison of integers of different signs: 'int' and 'const unsigned int' [-Wsign-compare]
|   136 |     bool gen_invalid = m_rng.randrange(100) < invalid_rate;
|       |                        ~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~
|   137 |     bool gen_fork = m_rng.randrange(100) < branch_rate;
|       |                     ~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~
| 2 warnings generated.

Commit:
test: refactor: Accept any RandomNumberGenerator in RandMoney

Accepting any Rng in RandMoney makes tests more flexible to use a
different Rng. Also, passing in the Rng clarifies the call sites, so
that they all use g_rand_ctx explicitly and consistently.

ismaelsadeeq reviewed (commented):
Concept ACK

marcofleon reviewed (commented):
Lgtm ACK fa295215f6fa3b85b0387511920f75eeb3e12b58. Reviewed the code and ran `make check`. Also ran the suggested test:
`RANDOM_CTX_SEED=z ./src/test/test_bitcoin --log_level=all --run_test=timeoffsets_tests/timeoffsets_warning -- -printtoconsole=1 | grep RANDOM_CTX_SEED`

marcofleon commented on src/test/util/random.h:
```diff
@@ -27,34 +27,10 @@ enum class SeedRand {
-static inline CAmount InsecureRandMoneyAmount()
-{
-    return static_cast<CAmount>(InsecureRandRange(MAX_MONEY + 1));
+    return CAmount{rng.randrange(MAX_MONEY + 1)};
```
Just for my own understanding, the advantage here is now that any rng in a test can be used when calling RandMoney?

marcofleon replied:
Ah got it, makes sense. I need to check the individual commits. Thanks.

ryanofsky reviewed (approved):
Code review ACK fa295215f6fa3b85b0387511920f75eeb3e12b58 but I think it would be nicer to add `FastRandomContext& m_rng{g_rand_ctx}` to `BasicTestingSetup` and have most test code use `m_rng` instead of `g_rand_ctx` directly.

I think this would be nicer mostly as a matter of style since most test code now doesn't reference global variables directly, while this PR is adding hundreds of references. But it could also help down the road, for example if we wanted to remove the global, or run test cases deterministically in parallel. Current change looks ok to me though if this suggestion is too much work.

Commit:
test: Add m_rng alias for the global random context

The two names point to the same object, but having the reference now
allows easier removal of the global in the future.

ryanofsky committed:
test: refactor: Give unit test functions access to test state

Add unit test subclasses as needed so unit test functions that need to access
members like m_rng can reference it directly.

ryanofsky committed:
test: refactor: Pass rng parameters to test functions

Add FastRandomContext parameter to the utility function
AddTestCoin(), and a few local test functions and classes.

ryanofsky committed:
test: Remove FastRandomContext global

Drop g_insecure_rand_ctx

ryanofsky commented:
Code review ACK facc85b766dec2481384c878ccbf1696a8386e10. But I think we can eliminate the global now, and I think would be good to do it in this PR to changing affected tests more than once. I posted a branch https://github.com/ryanofsky/bitcoin/commits/pr/rng which I hope you can steal. It has the following commits:

- a334a9b9bd0d8ee520f9e2769534625a5d147705 test: Add m_rng alias for the global random context
- c38390fb43e40913afa2072c7986ea80da1d4d38 test: refactor: Make unsigned promotion explicit
- 51712d8ee0cc8b03a692f08d8a39b0e7a7fe8609 test: refactor: Give unit test functions access to test state
- ce7ef02cee703e985554c7ced196684f33f96fb9 test: refactor: Pass rng parameters to test functions
- 678c3df2e3a2c7366646db3ec0deb7cad0ede107 test: refactor: Accept any RandomNumberGenerator in RandMoney
- c1c384e57619d9e78c7af9e3ee098f3f25fd5479 scripted-diff: [test] Use g_rng/m_rng directly
- 8468112a8262054d71caaa90764bfa88548a82b8 test: Remove FastRandomContext global

maflcko reviewed (commented):
Sure, pushed your branch. A skim looks good. I left a nit and will do a real review later.

ryanofsky commented on src/test/util/random.h:
```diff
@@ -25,7 +25,7 @@ enum class SeedRand {
 
 /** Seed the RNG for testing. This affects all randomness, except GetStrongRandBytes(). */
-void SeedRandomForTest(SeedRand seed = SeedRand::SEED);
+void SeedRandomForTest(FastRandomContext& rng, SeedRand seed = SeedRand::SEED);
```
> nit in [bb67ca0](https://github.com/bitcoin/bitcoin/commit/bb67ca09c88a46b99bf83e937fa761f0e25cba37): Not sure about this change. It seems overly restrictive to require a `FastRandomContext&` here, when a caller may not want to pass anything at all, or a different rng.

Makes sense. I hadn't really looked at this function closely and was just trying to keep behavior unchanged. Maybe it would be good to change signature to something like:

```c++
SeedRandomForTest(SeedRand seed = SeedRand::SEED, FastRandomContext* rng = nullptr)
```

and change the last line of the function to

```c++
if (rng) rng->Reseed(GetRandHash());
```

so it could still be used conveniently to seed m_rng. Or maybe the parameter could just be dropped like the global `g_insecure_rand_ctx` is dropped, and `m_rng` could be managed separately.

ryanofsky reviewed (approved):
Code review ACK f1c1b37d2247eed49156c0467daa68aa38497bb8. Current version of this looks identical to the branch I pushed, except it is rebased and whitespace in the scripted diff was adjusted.

Commit:
test: refactor: Move g_insecure_rand_ctx.Reseed out of the helper that calls MakeRandDeterministicDANGEROUS

The global g_insecure_rand_ctx will be removed in the future, so
removing it from this helper is useful.

Also, tying the two concepts of the global internal RNGState and the
global test-only rng context is a bit confusing, because tests can
simply use the m_rng, if it exists. Also, tests may seed more than one
random context, or none at all, or a random context of a different type.

Fix all issues by moving the Reseed call to the two places where it is
used.

ryanofsky commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -102,12 +102,6 @@ void ResetCoverageCounters() {}
-    // randomness during the fuzz test, except:
-    // - GetStrongRandBytes(), which is used for the creation of private key material.
-    // - Creating a BasicTestingSetup or derived class will switch to a random seed.
-    SeedRandomForTest(g_insecure_rand_ctx, SeedRand::ZEROS);
```
re: https://github.com/bitcoin/bitcoin/pull/30571#discussion_r1719563783

> q in [f1c1b37](https://github.com/bitcoin/bitcoin/commit/f1c1b37d2247eed49156c0467daa68aa38497bb8): Can you explain why this change is correct?

Of course it is not correct. I only noticed that this function was updating  `g_insecure_rand_ctx` and didn't notice it was updating global rng state. Thanks for fixing!

ryanofsky reviewed (approved):
Code review ACK 164732369fc5ecf4c7705136a2de884419023b5a. New commit replacing SeedRandomForTest is very nice and clarifies the code.

hodlinator reviewed (commented):
Concept ACK 164732369fc5ecf4c7705136a2de884419023b5a

ryanofsky reviewed (approved):
Code review ACK 3836bf1e0a207a6cb659053404bcba3a44cf7bb2

Just tweaked cuckoocache_tests since last review

marcofleon commented:
Code review ACK 3836bf1e0a207a6cb659053404bcba3a44cf7bb2

marcofleon reviewed (commented):
Code review ACK 3836bf1e0a207a6cb659053404bcba3a44cf7bb2

hodlinator commented on src/test/prevector_tests.cpp:
```diff
@@ -208,84 +208,85 @@ class prevector_tester {
+        SeedRandomStateForTest(SeedRand::SEED);
+        rng.Reseed(GetRandHash());
+        rand_seed = rng.rand256();
         rand_cache.Reseed(rand_seed);
```
`rand_cache` is dead code since 16329224e70d0525208f6b0ba00c5e1531a4f5ea. Maybe remove it in an early clean-up commit?

hodlinator commented on src/test/util/setup_common.h:
```diff
@@ -64,6 +65,14 @@ struct BasicTestingSetup {
+    void SeedRandomForTest(SeedRand seed = SeedRand::SEED)
+    {
+        SeedRandomStateForTest(seed);
+        m_rng.Reseed(GetRandHash());
```
I understand this was the behavior in the past too, but should we do something else with `m_rng` if the function was called with `SeedRand::ZEROS`?

hodlinator commented on src/test/orphanage_tests.cpp:
```diff
@@ -29,14 +31,16 @@ class TxOrphanageTest : public TxOrphanage
             it = m_orphans.begin();
         return it->second.tx;
     }
+
+    FastRandomContext& m_rng;
```
Data-members should be private by default:
```diff
--- a/src/test/orphanage_tests.cpp
+++ b/src/test/orphanage_tests.cpp
@@ -20,6 +20,8 @@ BOOST_FIXTURE_TEST_SUITE(orphanage_tests, TestingSetup)
 
 class TxOrphanageTest : public TxOrphanage
 {
+    FastRandomContext& m_rng;
+
 public:
     TxOrphanageTest(FastRandomContext& rng) : m_rng{rng} {}
 
@@ -36,8 +38,6 @@ public:
             it = m_orphans.begin();
         return it->second.tx;
     }
-
-    FastRandomContext& m_rng;
 };
 
 static void MakeNewKeyWithFastRandomContext(CKey& key, FastRandomContext& rand_ctx)
```
Same thing with `CPartialMerkleTreeTester`.

hodlinator replied:
Defaulting to private means that should something outside the class start needing access to it, that PR would need to expose it, thereby highlighting to reviewers what is actually happening.

I think it's a good principle even if this is test code, it is new in this PR after all. Maybe if you re-touch?

hodlinator commented on src/test/util/random.h:
```diff
@@ -11,50 +11,18 @@
 
 #include <cstdint>
```
nit: No longer needed:
```diff
--- a/src/test/util/random.h
+++ b/src/test/util/random.h
@@ -7,9 +7,6 @@
 
 #include <consensus/amount.h>
 #include <random.h>
-#include <uint256.h>
-
-#include <cstdint>
 
 enum class SeedRand {
     ZEROS, //!< Seed with a compile time constant of zeros
```

Commit:
test: Correct the random seed log on a prevector test failure

rand_cache is unused since commit
16329224e70d0525208f6b0ba00c5e1531a4f5ea, so it can be removed

rand_seed is wrong since commit
022cf47dd7ef8f46e32a184e84f94d1e9f3a495c, because it is no longer
printing the seed that was used to seed the global random context in
tests. Instead, it prints a (random-ish) value derived from the global
random context via InsecureRand256().

Finally, the for loop creating new prevector_tester objects will always
use the same seed since commit fae43a97ca947cd0802392e9bb86d9d0572c0fba,
because repeated calls to SeedInsecureRand/SeedRandomForTest will always
reseed the global with the same "static const" seed.

Fix all issues by
* removing the unused rand_cache,
* removing the call to SeedRandomForTest which restored the same seed on
  every call in the process, and
* Reseeding the global random context with the (random-ish) rand_seed.

ryanofsky commented on src/test/prevector_tests.cpp:
```diff
@@ -205,13 +202,11 @@ class prevector_tester {
 
     ~prevector_tester() {
-        BOOST_CHECK_MESSAGE(passed, "insecure_rand: " + rand_seed.ToString());
+        BOOST_CHECK_MESSAGE(passed, "The random seed was logged at the start of the test");
```
In commit "test: Correct the random seed log on a prevector test failure" (faa2608e6c3e1763d0991443c3ce6d9bb1466da8)

I think it would be nice to just fix the bug and print the seed when the test fails rather than requiring looking up the seed in logs. It seems like we could easily support this by adding `return seed` as the last line of `SeedRandomForTest` and this would be helpful for debugging tests that only fail with certain seeds, because they could be run without always enabling logging.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30571#discussion_r1725675521

That all sounds good but I think I'm proposing a smaller not bigger change in behavior by just fixing the bug instead of trying to do something grander. The change based on master is:

```diff
--- a/src/test/prevector_tests.cpp
+++ b/src/test/prevector_tests.cpp
@@ -209,9 +209,8 @@ public:
     }
 
     prevector_tester() {
-        SeedRandomForTest();
-        rand_seed = InsecureRand256();
-        rand_cache.Reseed(rand_seed);
+        rand_seed = SeedRandomForTest();
+        rand_cache.Reseed(InsecureRand256());
     }
 };
 
--- a/src/test/util/random.cpp
+++ b/src/test/util/random.cpp
@@ -15,7 +15,7 @@ FastRandomContext g_insecure_rand_ctx;
 
 extern void MakeRandDeterministicDANGEROUS(const uint256& seed) noexcept;
 
-void SeedRandomForTest(SeedRand seedtype)
+uint256 SeedRandomForTest(SeedRand seedtype)
 {
     static const std::string RANDOM_CTX_SEED{"RANDOM_CTX_SEED"};
 
@@ -35,4 +35,5 @@ void SeedRandomForTest(SeedRand seedtype)
     LogPrintf("%s: Setting random seed for current tests to %s=%s\n", __func__, RANDOM_CTX_SEED, seed.GetHex());
     MakeRandDeterministicDANGEROUS(seed);
     g_insecure_rand_ctx.Reseed(GetRandHash());
+    return seed;
 }
--- a/src/test/util/random.h
+++ b/src/test/util/random.h
@@ -25,7 +25,7 @@ enum class SeedRand {
 };
 
 /** Seed the RNG for testing. This affects all randomness, except GetStrongRandBytes(). */
-void SeedRandomForTest(SeedRand seed = SeedRand::SEED);
+uint256 SeedRandomForTest(SeedRand seed = SeedRand::SEED);
 
 static inline uint32_t InsecureRand32()
 {

```

This would change the test less than the current PR, provide nicer behavior, and help out with future changes that make the seed more accessible, by returning the value from the function which set it.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30571#discussion_r1725758066

> Starting with commit https://github.com/bitcoin/bitcoin/commit/fae43a97ca947cd0802392e9bb86d9d0572c0fba, the seed is static const per process

Wow, I didn't notice `ctx_seed` was statically caching a random value before this so I've been misunderstanding what the `SeedRandomForTest` function has been doing the whole time. I think this could be made clearer by:

- Renaming `SeedRand::SEED` to `SeedRand::FIXED_SEED` so it is clearer `SeedRandomForTest` sets a fixed value, not one that can change between calls.
- Dropping the `SeedRandomForTest` default argument value so if a test resets to the fixed seed the calls look like `SeedRandomForTest(SeedRand::FIXED_SEED)` which tells you a fixed value is being set, as opposed to `SeedRandomForTest()` which makes it look like it could be setting a new random seed.
- Adding comments that better document what the seed functions and enum values do.

Would suggest:

```diff
--- a/src/test/util/random.h
+++ b/src/test/util/random.h
@@ -13,10 +13,13 @@
 
 enum class SeedRand {
     ZEROS, //!< Seed with a compile time constant of zeros
-    SEED,  //!< Use (and report) random seed from environment, or a (truly) random one.
+    FIXED_SEED, //!< Seed with a fixed value that never changes over the
+                //!< lifetime of this process. The seed is read from the
+                //!< RANDOM_CTX_SEED environment variable if set, otherwise
+                //!< generated randomly once, saved, and reused.
 };
 
-/** Seed the internal global RNG state for testing. This affects all randomness, except GetStrongRandBytes(). */
+/** Seed the global RNG state for testing and log the seed value. This affects all randomness, except GetStrongRandBytes(). */
 void SeedRandomStateForTest(SeedRand seed);
 
 template <RandomNumberGenerator Rng>
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -67,8 +67,8 @@ struct BasicTestingSetup {
     node::NodeContext m_node; // keep as first member to be destructed last
 
     FastRandomContext m_rng;
-    /** Seed the internal global RNG state and m_rng for testing. This affects all randomness, except GetStrongRandBytes(). */
-    void SeedRandomForTest(SeedRand seed = SeedRand::SEED)
+    /** Seed the global RNG state and m_rng for testing and log the seed value. This affects all randomness, except GetStrongRandBytes(). */
+    void SeedRandomForTest(SeedRand seed)
     {
         SeedRandomStateForTest(seed);
         m_rng.Reseed(GetRandHash());

```

ryanofsky reviewed (approved):
Code review ACK b6b2b38aa77162420babb34e01dd510512411abf. Just simplified prevector_tests and removed misleading BOOST_CHECK_MESSAGE print since last review.

hodlinator reviewed (commented):
Reviewed 38b03bc3af9fbf483b5a29a54a5ffb36c67b6b41 (still aiming to do some more testing).

> Thanks for checking the includes in the tests @hodlinator! I think I'll leave this as-is for now, because having leftover unused includes in tests should be harmless, apart from possibly being confusing, or needing more compile time. I have a follow-up to use iwyu (and enforce it), so that reviewers can use the re-review cycles to clean up includes for reviewing the code itself. However, this will be a follow-up, so I'll leave this pull request as-is for now.

Agree it's low-value cycles to [check #includes/forward declarations](https://github.com/bitcoin/bitcoin/pull/30571#pullrequestreview-2250158253), but until we start using iwyu everywhere, it felt fitting as you are doing major surgery in & around **test/util/random.h** in this case.

Great that my discovery helped spawn your additional fixes to **prevector_tests.cpp**!

ryanofsky reviewed (approved):
Code review ACK 38b03bc3af9fbf483b5a29a54a5ffb36c67b6b41. Just another random seed bug in the prevector tests fixed since last review!

hodlinator reviewed (approved):
ACK 948238a683b6c99f4e91114aa75680c6c2d73714

Gets rid of `g_insecure_rand_ctx`-global state along with global functions using it, in favor of local `m_rng` contexts.

Tested adding..
```C++
BOOST_CHECK_EQUAL(m_rng.rand64(), 0);
```
..to one of the tests and verified that the random number emitted varied when re-running the test, and became constant when called repeatedly with `RANDOM_CTX_SEED` env var set, and different constant when modifying `RANDOM_CTX_SEED` value.

(Also uncovered #30696 while testing).

#### fae7e3791c9ed8053166773fcfb583ad19d006dd "test: Correct the random seed log on a prevector test failure"

Good addition to the PR!

nit: The tense in the commit message feels off,
"rand_cache is unused since" -> "rand_cache was unused since"
"rand_seed is wrong" -> "rand_seed was wrong"
etc

ryanofsky reviewed (approved):
Code review ACK 948238a683b6c99f4e91114aa75680c6c2d73714. Only changes since last review were changing a comments a little bit.

marcofleon commented:
In the process of reviewing now.

marcofleon reviewed (approved):
Code review ACK 948238a683b6c99f4e91114aa75680c6c2d73714. Only changes since my last review are the improvements in `prevector_tests`.
